File: src/main/java/org/moinex/repositories/wallettransaction/WalletTransactionRepository.java
/*
 * Filename: WalletTransactionRepository.java
 * Created on: August 31, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.repositories.wallettransaction;

import java.util.List;
import java.util.Optional;
import org.moinex.entities.wallettransaction.Wallet;
import org.moinex.entities.wallettransaction.WalletTransaction;
import org.moinex.util.enums.TransactionType;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

/**
 * Repository for the WalletTransaction entity
 *
 * This repository provides methods to query the database for WalletTransaction
 *
 * Each method to get transactions has a version that returns only transactions
 * that have a category that is not archived
 */
@Repository
public interface WalletTransactionRepository
    extends JpaRepository<WalletTransaction, Long> {

    /**
     * Get all transactions where both the category and wallet are not archived
     * @return A list with all transactions
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedTransactions();

    /**
     * Get all income transactions
     * @return A list with all income transactions
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.type = 'INCOME' "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findIncomeTransactions();

    /**
     * Get all income transactions where both the category and wallet are not archived
     * @return A list with all income transactions
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.type = 'INCOME' "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedIncomeTransactions();

    /**
     * Get all expense transactions
     * @return A list with all expense transactions
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.type = 'EXPENSE' "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findExpenseTransactions();

    /**
     * Get all expense transactions where both the category and wallet are not archived
     * @return A list with all expense transactions
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.type = 'EXPENSE' "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedExpenseTransactions();

    /**
     * Get the all transactions by month and year
     * @param month The month
     * @param year The year
     * @return A list with the transactions by month and year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findTransactionsByMonth(@Param("month") Integer month, @Param("year") Integer year);

    /**
     * Get the all transactions by month and year where both the category and wallet are
     * not archived
     * @param month The month
     * @param year The year
     * @return A list with the transactions by month and year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedTransactionsByMonth(@Param("month") Integer month,
                                       @Param("year") Integer  year);

    /**
     * Get the all transactions by year
     * @param year The year
     * @return A list with the transactions by year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%Y', wt.date) = printf('%04d', :year) "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findTransactionsByYear(@Param("year") Integer year);

    /**
     * Get the all transactions by year where both the category and wallet are not
     * archived
     * @param year The year
     * @return A list with the transactions by year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%Y', wt.date) = printf('%04d', :year) "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedTransactionsByYear(@Param("year") Integer year);

    /**
     * Get the transactions by wallet and month
     * @param walletId The id of the wallet
     * @param month The month
     * @param year The year
     * @return A list with the transactions in the wallet by month
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.wallet.id = :walletId "
           + "AND strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findTransactionsByWalletAndMonth(@Param("walletId") Long walletId,
                                     @Param("month") Integer month,
                                     @Param("year") Integer  year);

    /**
     * Get the transactions by wallet and month where both the category and wallet are
     * not
     * @param walletId The id of the wallet
     * @param month The month
     * @param year The year
     * @return A list with the transactions in the wallet by month
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.wallet.id = :walletId "
           + "AND strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedTransactionsByWalletAndMonth(@Param("walletId") Long walletId,
                                                @Param("month") Integer month,
                                                @Param("year") Integer  year);

    /**
     * Get all transactions between two dates
     * @param startDate The start date
     * @param endDate The end date
     * @return A list with the transactions between the two dates
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.date >= :startDate "
           + "AND wt.date <= :endDate "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findTransactionsBetweenDates(@Param("startDate") String startDate,
                                 @Param("endDate") String   endDate);

    /**
     * Get all transactions between two dates where both the category and wallet are not
     * archived
     * @param startDate The start date
     * @param endDate The end date
     * @return A list with the transactions between the two dates
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.date >= :startDate "
           + "AND wt.date <= :endDate "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedTransactionsBetweenDates(@Param("startDate") String startDate,
                                            @Param("endDate") String   endDate);

    /**
     * Get the confirmed transactions by month and year
     * @param month The month
     * @param year The year
     * @return A list with the transactions in the wallet by month and year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "AND wt.status = 'CONFIRMED' "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findConfirmedTransactionsByMonth(@Param("month") Integer month,
                                     @Param("year") Integer  year);

    /**
     * Get the confirmed transactions by month and year where both the category and
     * wallet are not archived
     * @param month The month
     * @param year The year
     * @return A list with the transactions in the wallet by month and year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "AND wt.status = 'CONFIRMED' "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedConfirmedTransactionsByMonth(@Param("month") Integer month,
                                                @Param("year") Integer  year);

    /**
     * Get the pending transactions by month and year
     * @param month The month
     * @param year The year
     * @return A list with the transactions in the wallet by month and year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "AND wt.status = 'PENDING' "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findPendingTransactionsByMonth(@Param("month") Integer month,
                                   @Param("year") Integer  year);

    /**
     * Get the pending transactions by month and year where both the category and wallet
     * are not archived
     * @param month The month
     * @param year The year
     * @return A list with the transactions in the wallet by month and year
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE strftime('%m', wt.date) = printf('%02d', :month) "
           + "AND strftime('%Y', wt.date) = printf('%04d', :year) "
           + "AND wt.status = 'PENDING' "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedPendingTransactionsByMonth(@Param("month") Integer month,
                                              @Param("year") Integer  year);

    /**
     * Get the last n transactions of all wallets
     * @param pageable The pageable object
     * @return A list with the last n transactions of all wallets
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findLastTransactions(Pageable pageable);

    /**
     * Get the last n transactions of all wallets where both the category and wallet are
     * not archived
     * @param pageable The pageable object
     * @return A list with the last n transactions of all wallets
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedLastTransactions(Pageable pageable);

    /**
     * Get the last n transactions in a wallet
     * @param walletId The id of the wallet
     * @param pageable The pageable object
     * @return A list with the last n transactions in the wallet
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.wallet.id = :walletId "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findLastTransactionsByWallet(@Param("walletId") Long walletId, Pageable pageable);

    /**
     * Get the last n transactions in a wallet where both the category and wallet are
     * not archived
     * @param walletId The id of the wallet
     * @param pageable The pageable object
     * @return A list with the last n transactions in the wallet
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.wallet.id = :walletId "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findNonArchivedLastTransactionsByWallet(@Param("walletId") Long walletId,
                                            Pageable                pageable);

    /**
     * Get the date of the oldest transaction
     * @return The date of the oldest transaction
     */
    @Query("SELECT min(wt.date) "
           + "FROM WalletTransaction wt")
    String
    findOldestTransactionDate();

    /**
     * Get the date of the oldest transaction where both the category and wallet are not
     * archived
     * @return The date of the oldest transaction
     */
    @Query("SELECT min(wt.date) "
           + "FROM WalletTransaction wt "
           + "WHERE wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false")
    String
    findNonArchivedOldestTransactionDate();

    /**
     * Get the date of the newest transaction
     * @return The date of the newest transaction
     */
    @Query("SELECT max(wt.date) "
           + "FROM WalletTransaction wt")
    String
    findNewestTransactionDate();

    /**
     * Get the date of the newest transaction and the category is not archived
     * @return The date of the newest transaction
     */
    @Query("SELECT max(wt.date) "
           + "FROM WalletTransaction wt "
           + "WHERE wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false")
    String
    findNonArchivedNewestTransactionDate();

    /**
     * Get count of transactions by wallet
     * @param walletId The id of the wallet
     * @return The count of transactions in the wallet
     */
    @Query("SELECT count(wt) "
           + "FROM WalletTransaction wt "
           + "WHERE wt.wallet.id = :walletId")
    Long
    getTransactionCountByWallet(@Param("walletId") Long walletId);

    /**
     * Get count of transactions by wallet where both the category and wallet are not
     * archived
     * @param walletId The id of the wallet
     * @return The count of transactions in the wallet
     */
    @Query("SELECT count(wt) "
           + "FROM WalletTransaction wt "
           + "WHERE wt.wallet.id = :walletId "
           + "AND wt.category.isArchived = false "
           + "AND wt.wallet.isArchived = false")
    Long
    getCountNonArchivedTransactionsByWallet(@Param("walletId") Long walletId);

    /**
     * Get the wallet by transaction id
     * @param transactionId The id of the transaction
     */
    @Query("SELECT wt.wallet "
           + "FROM WalletTransaction wt "
           + "WHERE wt.id = :transactionId")
    Optional<Wallet>
    findWalletByTransactionId(@Param("transactionId") Long transactionId);

    /**
     * Check if a wallet exists by transaction id
     * @param transactionId The id of the transaction
     * @return True if the wallet exists, false otherwise
     */
    @Query(
        "SELECT COUNT(wt) > 0 FROM WalletTransaction wt WHERE wt.id = :transactionId")
    boolean
    existsWalletByTransactionId(@Param("transactionId") Long transactionId);

    /**
     * Get suggestions. Suggestions are transactions with distinct descriptions
     * and most recent date
     * @param transactionType The type of the transaction
     * @return A list with the suggestions
     */
    @Query("SELECT wt "
           + "FROM WalletTransaction wt "
           + "WHERE wt.type = :transactionType AND "
           + "wt.wallet.isArchived = false AND "
           + "wt.category.isArchived = false AND "
           + "wt.date = (SELECT max(wt2.date) "
           + "                  FROM WalletTransaction wt2 "
           + "                  WHERE wt2.wallet.isArchived = false AND "
           + "                        wt2.category.isArchived = false AND "
           + "                        wt2.description = wt.description) "
           + "ORDER BY wt.date DESC")
    List<WalletTransaction>
    findSuggestions(@Param("transactionType") TransactionType transactionType);
}

File: src/main/kotlin/org/moinex/repository/wallettransaction/WalletTransactionRepository.kt
/*
 * Filename: WalletTransactionRepository.kt (original filename: WalletTransactionRepository.java)
 * Created on: August 31, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 03/03/2026
 */

package org.moinex.repository.wallettransaction

import org.moinex.model.enums.WalletTransactionType
import org.moinex.model.wallettransaction.WalletTransaction
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.math.BigDecimal
import java.time.LocalDateTime

@Repository
interface WalletTransactionRepository : JpaRepository<WalletTransaction, Int> {
    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedTransactions(): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.type = 'INCOME' " +
            "ORDER BY wt.date DESC",
    )
    fun findIncomeTransactions(): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.type = 'EXPENSE' " +
            "ORDER BY wt.date DESC",
    )
    fun findExpenseTransactions(): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.type = :type " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findAllNonArchivedTransactionsByType(
        @Param("type") type: WalletTransactionType,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "ORDER BY wt.date DESC",
    )
    fun findTransactionsByMonth(
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedTransactionsByMonth(
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%Y', wt.date) = printf('%04d', :year) " +
            "ORDER BY wt.date DESC",
    )
    fun findTransactionsByYear(
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedTransactionsByYear(
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "AND strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "ORDER BY wt.date DESC",
    )
    fun findTransactionsByWalletAndMonth(
        @Param("walletId") walletId: Int,
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "AND strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedTransactionsByWalletAndMonth(
        @Param("walletId") walletId: Int,
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.date >= :startDate " +
            "AND wt.date <= :endDate " +
            "ORDER BY wt.date DESC",
    )
    fun findTransactionsBetweenDates(
        @Param("startDate") startDate: LocalDateTime,
        @Param("endDate") endDate: LocalDateTime,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.date >= :startDate " +
            "AND wt.date <= :endDate " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedTransactionsBetweenDates(
        @Param("startDate") startDate: LocalDateTime,
        @Param("endDate") endDate: LocalDateTime,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.status = 'CONFIRMED' " +
            "ORDER BY wt.date DESC",
    )
    fun findConfirmedTransactionsByMonth(
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.status = 'CONFIRMED' " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedConfirmedTransactionsByMonth(
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.status = 'PENDING' " +
            "ORDER BY wt.date DESC",
    )
    fun findPendingTransactionsByMonth(
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.status = 'PENDING' " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedPendingTransactionsByMonth(
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "ORDER BY wt.date DESC",
    )
    fun findLastTransactions(pageable: Pageable): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedLastTransactions(pageable: Pageable): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "ORDER BY wt.date DESC",
    )
    fun findLastTransactionsByWallet(
        @Param("walletId") walletId: Int,
        pageable: Pageable,
    ): List<WalletTransaction>

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedLastTransactionsByWallet(
        @Param("walletId") walletId: Int,
        pageable: Pageable,
    ): List<WalletTransaction>

    @Query(
        "SELECT min(wt.date) " +
            "FROM WalletTransaction wt",
    )
    fun findOldestTransactionDate(): LocalDateTime?

    @Query(
        "SELECT min(wt.date) " +
            "FROM WalletTransaction wt " +
            "WHERE wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false",
    )
    fun findNonArchivedOldestTransactionDate(): LocalDateTime?

    @Query(
        "SELECT max(wt.date) " +
            "FROM WalletTransaction wt",
    )
    fun findNewestTransactionDate(): LocalDateTime?

    @Query(
        "SELECT max(wt.date) " +
            "FROM WalletTransaction wt " +
            "WHERE wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false",
    )
    fun findNonArchivedNewestTransactionDate(): LocalDateTime?

    @Query(
        "SELECT count(wt) " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId",
    )
    fun getTransactionCountByWallet(
        @Param("walletId") walletId: Int,
    ): Int

    @Query(
        "SELECT count(wt) " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false",
    )
    fun getCountNonArchivedTransactionsByWallet(
        @Param("walletId") walletId: Int,
    ): Int

    @Query(
        "SELECT COUNT(wt) > 0 " +
            "FROM WalletTransaction wt " +
            "WHERE wt.id = :transactionId",
    )
    fun existsWalletByTransactionId(
        @Param("transactionId") transactionId: Int,
    ): Boolean

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.type = :walletTransactionType " +
            "AND wt.wallet.isArchived = false " +
            "AND wt.category.isArchived = false " +
            "AND wt.date = (SELECT max(wt2.date) " +
            "               FROM WalletTransaction wt2 " +
            "               WHERE wt2.wallet.isArchived = false " +
            "               AND wt2.category.isArchived = false " +
            "               AND wt2.description = wt.description) " +
            "ORDER BY wt.date DESC",
    )
    fun findSuggestions(
        @Param("walletTransactionType") walletTransactionType: WalletTransactionType,
    ): List<WalletTransaction>

    @Query(
        "SELECT COALESCE(SUM(wt.amount), 0) " +
            "FROM WalletTransaction wt " +
            "WHERE wt.category.id IN :categoryIds " +
            "AND wt.type = :walletTransactionType " +
            "AND wt.date BETWEEN :startDate AND :endDate",
    )
    fun sumAmountByCategoriesAndDateRange(
        @Param("categoryIds") categoryIds: List<Int>,
        @Param("walletTransactionType") walletTransactionType: WalletTransactionType,
        @Param("startDate") startDate: LocalDateTime,
        @Param("endDate") endDate: LocalDateTime,
    ): BigDecimal

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE strftime('%m', wt.date) = printf('%02d', :month) " +
            "AND strftime('%Y', wt.date) = printf('%04d', :year) " +
            "AND wt.includeInAnalysis = true " +
            "AND wt.category.isArchived = false " +
            "AND wt.wallet.isArchived = false " +
            "ORDER BY wt.date DESC",
    )
    fun findNonArchivedTransactionsByMonthForAnalysis(
        @Param("month") month: Int,
        @Param("year") year: Int,
    ): List<WalletTransaction>

    @Query(
        "SELECT COALESCE(SUM(wt.amount), 0) " +
            "FROM WalletTransaction wt " +
            "WHERE wt.category.id IN :categoryIds " +
            "AND wt.type = :walletTransactionType " +
            "AND wt.includeInAnalysis = true " +
            "AND wt.date BETWEEN :startDate AND :endDate",
    )
    fun getSumAmountByCategoriesAndDateBetweenForAnalysis(
        @Param("categoryIds") categoryIds: List<Int>,
        @Param("walletTransactionType") walletTransactionType: WalletTransactionType,
        @Param("startDate") startDate: LocalDateTime,
        @Param("endDate") endDate: LocalDateTime,
    ): BigDecimal

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "AND wt.date >= :date " +
            "ORDER BY wt.date ASC",
    )
    fun findTransactionsByWalletAfterDate(
        @Param("walletId") walletId: Int,
        @Param("date") date: LocalDateTime,
    ): List<WalletTransaction>

    @Query(
        "SELECT MIN(wt.date) " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "AND wt.status = 'CONFIRMED'",
    )
    fun findFirstTransactionDate(
        @Param("walletId") walletId: Int,
    ): LocalDateTime?

    @Query(
        "SELECT wt " +
            "FROM WalletTransaction wt " +
            "WHERE wt.wallet.id = :walletId " +
            "AND wt.date <= :endDate " +
            "AND wt.status = 'CONFIRMED' " +
            "ORDER BY wt.date ASC",
    )
    fun findAllConfirmedTransactionsUpToDate(
        @Param("walletId") walletId: Int,
        @Param("endDate") endDate: LocalDateTime,
    ): List<WalletTransaction>

    @Query(
        "SELECT count(t) " +
            "FROM WalletTransaction t " +
            "WHERE t.category.id = :categoryId",
    )
    fun getTransactionCountByCategory(
        @Param("categoryId") categoryId: Int,
    ): Int
}
